semantics/src/main/kotlin/com/strumenta/kolasu/semantics/identifier/provider/declarative/DeclarativeIdentifierProvider.kt

```diff
@@ -0,0 +1,60 @@
+package com.strumenta.kolasu.semantics.identifier.provider.declarative
+
+import com.strumenta.kolasu.model.Node
+import com.strumenta.kolasu.semantics.identifier.provider.IdentifierProvider
+import kotlin.reflect.KClass
+import kotlin.reflect.full.isSubclassOf
+import kotlin.reflect.full.isSuperclassOf
+
+inline fun <reified NodeTy : Node> identifierFor(
+    noinline specification: (DeclarativeIdentifierProviderRuleArguments<NodeTy>) -> String?
+): DeclarativeIdentifierProviderRule<NodeTy> {
+    return DeclarativeIdentifierProviderRule(NodeTy::class, specification)
+}
+
+open class DeclarativeIdentifierProvider(
+    vararg rules: DeclarativeIdentifierProviderRule<out Node>
+) : IdentifierProvider {
+    private val rules: List<DeclarativeIdentifierProviderRule<out Node>> = rules.sorted()
+
+    override fun <NodeTy : Node> getIdentifierFor(
+        node: NodeTy,
+        typedAs: KClass<in NodeTy>?
+    ): String? {
+        return this.rules
+            .firstOrNull { it.canBeInvokedWith(typedAs ?: node::class) }
+            ?.invoke(this, node)
+    }
+}
+
+class DeclarativeIdentifierProviderRule<NodeTy : Node>(
+    private val nodeType: KClass<NodeTy>,
+    private val specification: (DeclarativeIdentifierProviderRuleArguments<NodeTy>) -> String?
+) : Comparable<DeclarativeIdentifierProviderRule<*>>, (IdentifierProvider, Node) -> String? {
+    override fun invoke(
+        identifierProvider: IdentifierProvider,
+        node: Node
+    ): String? {
+        @Suppress("UNCHECKED_CAST")
+        return (node as? NodeTy)
+            ?.let { DeclarativeIdentifierProviderRuleArguments(it, identifierProvider) }
+            ?.let(specification)
+    }
+
+    fun canBeInvokedWith(withType: KClass<*>): Boolean {
+        return this.nodeType.isSuperclassOf(withType)
+    }
+
+    override fun compareTo(other: DeclarativeIdentifierProviderRule<*>): Int {
+        return when {
+            this.nodeType.isSuperclassOf(other.nodeType) -> 1
+            this.nodeType.isSubclassOf(other.nodeType) -> -1
+            else -> (this.nodeType.qualifiedName ?: "") compareTo (other.nodeType.qualifiedName ?: "")
+        }
+    }
+}
+
+data class DeclarativeIdentifierProviderRuleArguments<NodeTy : Node>(
```

Comment:
What is the role of this class?

Reply:
This is now renamed to `DeclarativeIdentifierProviderRuleContext` and it contains everything that can be accessed while defining an identifier provider rule.
```kotlin
object MyIdentifierProvider: DeclarativeIdentifierProvider(
    identifierFor(SomeNode::class) {
        // if nothing is specified, the context will be accessible through the it variable    
    },
    identifierFor(SomeNode::class) { (node: SomeNode, identifiers: IdentifierProvider) ->
        // node is the current node we are defining the identifier for
        // identifierProvider can be used to compute identifiers for other nodes
        // It could be used to compute the identifier of the container node, for example.
    }
)
```
This mechanism allows us to reuse the identifier inside its configuration, we would not be able to use `MyIdentifierProvider` otherwise - at least not using `val` or `object` definitions. We would probably need to subclass `DeclarativeIdentifierProvider` and this is considered as anti-pattern of our symbol resolution APIs.